File: whiteboard/app/src/main/java/com/ditto/whiteboard/protocol/SnapshotTransfer.kt
package com.ditto.whiteboard.protocol

import java.util.UUID

/** Hard protocol bounds. They are enforced before any peer-controlled allocation occurs. */
internal const val MAX_TRANSFER_BYTES: Long = 16L * 1024 * 1024
internal const val MAX_TRANSFER_CHUNKS: Int = 4_096
internal const val MAX_TRANSFER_ID_LENGTH: Int = 128
internal const val SHA_256_BYTE_COUNT: Int = 32

/**
 * A board snapshot split into size-bounded chunks for the Reliable `wb_state` stream, used when a
 * peer joins late and must catch up. [SnapshotTransfer.create] chunks the payload and records a
 * SHA-256 [digest]; the receiving [SnapshotAssembler] reassembles the chunks and verifies both the
 * byte count and the digest before the snapshot is accepted, so a corrupt or partial transfer is
 * rejected rather than merged.
 */
data class SnapshotTransfer(
  val id: String,
  val bytes: ByteArray,
  val digest: ByteArray,
  val chunks: List<ByteArray>,
) {
  companion object {
    fun create(bytes: ByteArray, maxChunkSize: Int, id: String = UUID.randomUUID().toString()): SnapshotTransfer {
      require(maxChunkSize > 0)
      require(bytes.isNotEmpty()) { "Transfer payload must not be empty" }
      require(bytes.size.toLong() <= MAX_TRANSFER_BYTES) {
        "Transfer exceeds the ${MAX_TRANSFER_BYTES / (1024 * 1024)} MiB protocol limit"
      }
      require(id.isNotBlank() && id.length <= MAX_TRANSFER_ID_LENGTH)
      val chunkCount = (bytes.size.toLong() + maxChunkSize - 1L) / maxChunkSize
      require(chunkCount <= MAX_TRANSFER_CHUNKS) { "Transfer requires too many chunks" }
      val chunks = buildList(chunkCount.toInt()) {
        var offset = 0
        while (offset < bytes.size) {
          val end = (offset + maxChunkSize).coerceAtMost(bytes.size)
          add(bytes.copyOfRange(offset, end))
          offset = end
        }
      }
      return SnapshotTransfer(id, bytes, WhiteboardProtocol.sha256(bytes), chunks)
    }
  }
}

sealed interface SnapshotAssemblyResult {
  data object Pending : SnapshotAssemblyResult
  data class Complete(val bytes: ByteArray) : SnapshotAssemblyResult
  data class Rejected(val reason: String) : SnapshotAssemblyResult
}

class SnapshotAssembler(
  private val transferId: String,
  private val chunkCount: Int,
  private val byteCount: Long,
  private val expectedDigest: ByteArray,
) {
  private val chunks = arrayOfNulls<ByteArray>(chunkCount)
  private var receivedByteCount = 0L
  private var sealed = false

  init {
    require(transferId.isNotBlank() && transferId.length <= MAX_TRANSFER_ID_LENGTH)
    require(chunkCount in 1..MAX_TRANSFER_CHUNKS)
    require(byteCount in 1..MAX_TRANSFER_BYTES)
    require(expectedDigest.size == SHA_256_BYTE_COUNT)
  }

  @Synchronized
  fun add(index: Int, bytes: ByteArray): SnapshotAssemblyResult {
    if (sealed) return SnapshotAssemblyResult.Rejected("Transfer already ended")
    if (index !in chunks.indices) return SnapshotAssemblyResult.Rejected("Chunk index out of range")
    val previousSize = chunks[index]?.size ?: 0
    val nextByteCount = receivedByteCount - previousSize + bytes.size
    if (nextByteCount > byteCount || nextByteCount > MAX_TRANSFER_BYTES) {
      return SnapshotAssemblyResult.Rejected("Transfer exceeds its declared byte count")
    }
    chunks[index] = bytes.copyOf()
    receivedByteCount = nextByteCount
    return SnapshotAssemblyResult.Pending
  }

  @Synchronized
  fun finish(id: String): SnapshotAssemblyResult {
    sealed = true
    if (id != transferId) return SnapshotAssemblyResult.Rejected("Transfer ID mismatch")
    if (chunks.any { it == null }) return SnapshotAssemblyResult.Rejected("Snapshot is incomplete")
    if (receivedByteCount != byteCount) return SnapshotAssemblyResult.Rejected("Snapshot size mismatch")
    val bytes = ByteArray(byteCount.toInt())
    var offset = 0
    chunks.filterNotNull().forEach { chunk ->
      chunk.copyInto(bytes, destinationOffset = offset)
      offset += chunk.size
    }
    if (!WhiteboardProtocol.sha256(bytes).contentEquals(expectedDigest)) {
      return SnapshotAssemblyResult.Rejected("Snapshot digest mismatch")
    }
    return SnapshotAssemblyResult.Complete(bytes)
  }
}

File: whiteboard-apple/Packages/WhiteboardCore/Sources/WhiteboardCore/Protocol/SnapshotTransfer.swift
import Foundation

public let maxTransferBytes: Int64 = 16 * 1024 * 1024
public let maxTransferChunks: Int = 4_096
public let maxTransferIdLength: Int = 128
public let sha256ByteCount: Int = 32

public struct SnapshotTransfer: Equatable, Sendable {
  public let id: String
  public let bytes: Data
  public let digest: Data
  public let chunks: [Data]

  public static func create(
    _ bytes: Data,
    maxChunkSize: Int,
    id: String = UUID().uuidString.lowercased()
  ) throws -> SnapshotTransfer {
    func require(_ condition: Bool, _ message: String = "") throws {
      if !condition { throw WhiteboardCoreError.requirementFailed(message) }
    }
    try require(maxChunkSize > 0)
    try require(!bytes.isEmpty, "Transfer payload must not be empty")
    try require(
      Int64(bytes.count) <= maxTransferBytes,
      "Transfer exceeds the \(maxTransferBytes / (1024 * 1024)) MiB protocol limit"
    )
    try require(!id.isBlankKotlin && id.utf16.count <= maxTransferIdLength)
    let chunkCount = (Int64(bytes.count) + Int64(maxChunkSize) - 1) / Int64(maxChunkSize)
    try require(chunkCount <= Int64(maxTransferChunks), "Transfer requires too many chunks")
    var chunks: [Data] = []
    chunks.reserveCapacity(Int(chunkCount))
    var offset = 0
    while offset < bytes.count {
      let end = min(offset + maxChunkSize, bytes.count)
      chunks.append(bytes[offset..<end])
      offset = end
    }
    return SnapshotTransfer(id: id, bytes: bytes, digest: WhiteboardProtocol.sha256(bytes), chunks: chunks)
  }

  public init(id: String, bytes: Data, digest: Data, chunks: [Data]) {
    self.id = id
    self.bytes = bytes
    self.digest = digest
    self.chunks = chunks
  }
}

public enum SnapshotAssemblyResult: Equatable, Sendable {
  case pending
  case complete(Data)
  case rejected(String)
}

public struct SnapshotAssembler: Sendable {
  private let transferId: String
  private let chunkCount: Int
  private let byteCount: Int64
  private let expectedDigest: Data
  private var chunks: [Data?]
  private var receivedByteCount: Int64 = 0
  private var sealed = false

  public init(transferId: String, chunkCount: Int, byteCount: Int64, expectedDigest: Data) throws {
    func require(_ condition: Bool, _ message: String = "") throws {
      if !condition { throw WhiteboardCoreError.requirementFailed(message) }
    }
    try require(!transferId.isBlankKotlin && transferId.utf16.count <= maxTransferIdLength)
    try require((1...maxTransferChunks).contains(chunkCount))
    try require((1...maxTransferBytes).contains(byteCount))
    try require(expectedDigest.count == sha256ByteCount)
    self.transferId = transferId
    self.chunkCount = chunkCount
    self.byteCount = byteCount
    self.expectedDigest = expectedDigest
    self.chunks = [Data?](repeating: nil, count: chunkCount)
  }

  public mutating func add(index: Int, bytes: Data) -> SnapshotAssemblyResult {
    if sealed { return .rejected("Transfer already ended") }
    guard chunks.indices.contains(index) else { return .rejected("Chunk index out of range") }
    let previousSize = chunks[index]?.count ?? 0
    let nextByteCount = receivedByteCount - Int64(previousSize) + Int64(bytes.count)
    if nextByteCount > byteCount || nextByteCount > maxTransferBytes {
      return .rejected("Transfer exceeds its declared byte count")
    }
    chunks[index] = bytes
    receivedByteCount = nextByteCount
    return .pending
  }

  public mutating func finish(_ id: String) -> SnapshotAssemblyResult {
    sealed = true
    if id != transferId { return .rejected("Transfer ID mismatch") }
    if chunks.contains(where: { $0 == nil }) { return .rejected("Snapshot is incomplete") }
    if receivedByteCount != byteCount { return .rejected("Snapshot size mismatch") }
    var bytes = Data()
    bytes.reserveCapacity(Int(byteCount))
    for chunk in chunks {
      if let chunk { bytes.append(chunk) }
    }
    if WhiteboardProtocol.sha256(bytes) != expectedDigest {
      return .rejected("Snapshot digest mismatch")
    }
    return .complete(bytes)
  }
}
